Stop asserting positive array size

xbauch opened the pull request:
when generating non-deterministic arrays. Zero-sized arrays are (unfortunately)
legal C construct (as for example the regression test shows).

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

hannes-steffenhagen-diffblue commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -163,7 +163,6 @@ void symbol_factoryt::gen_nondet_array_init(
   const auto &size = array_type.size();
   PRECONDITION(size.id() == ID_constant);
   auto const array_size = numeric_cast_v<size_t>(to_constant_expr(size));
-  DATA_INVARIANT(array_size > 0, "Arrays should have positive size");
```
I wasn't properly taking this situation into account when writing this.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4ad15b4090c57f821c0c00a38a30a166b9cfc9a2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/119815198

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: b6a8ea39aebbc2295356c4c25c1fd614b2ca2653).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/119982431
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

petr-bauch committed:
Stop asserting positive array size

when generating non-deterministic arrays. Zero-sized arrays are (unfortunately)
legal C construct (as for example the regression test shows).

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 80b914ed2797423e50b8f02c0c103eb47f660c79).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/119990011

kroening commented on src/solvers/smt2/smt2_conv.cpp:
```diff
@@ -2695,8 +2695,8 @@ void smt2_convt::flatten_array(const exprt &expr)
-  CHECK_RETURN_WITH_DIAGNOSTICS(size != 0, "can't convert zero-sized array");
-
+  if(size == 0)
+    return;
```
This is suspicious -- this produces an SMT file that might now parse.
Did you want to make that a precondition, and then check at the call site?

xbauch replied:
I reverted here to the original `CHECK_RETURN` and pushed the check to the above `convert_struct`.

kroening reviewed (changes requested):
The SMT-Lib2 file would not parse!

xbauch commented:
> The SMT-Lib2 file would not parse!

It should now (I had to fix another, seemingly unrelated, bug in the last commit). I added a test that runs `--smt2`.

xbauch commented on src/solvers/smt2/smt2_conv.cpp:
```diff
@@ -2672,7 +2672,14 @@ void smt2_convt::convert_struct(const struct_exprt &expr)
+        {
+          const array_typet &array_type = to_array_type(op.type());
+          const auto &size_expr = array_type.size();
+          PRECONDITION(size_expr.id() == ID_constant);
```
Not really a precondition, check-return?

smowton reviewed (approved):
Looks sensible

petr-bauch committed:
Also skip flattening zero-sized arrays

when converting to SMT2 queries.

xbauch commented:
@kroening `--smt2` test removed (z3 is unavailable on CI).

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 30923ab6a23ef7f7f0f9e9e6c099e167058daf11).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/121222947

tautschnig commented:
@kroening ?

kroening commented on src/solvers/smt2/smt2_conv.cpp:
```diff
@@ -2672,7 +2672,14 @@ void smt2_convt::convert_struct(const struct_exprt &expr)
+          const array_typet &array_type = to_array_type(op.type());
+          const auto &size_expr = array_type.size();
+          CHECK_RETURN(size_expr.id() == ID_constant);
+
```
This prevents using arrays with variable size.

kroening commented on src/solvers/smt2/smt2_conv.cpp:
```diff
@@ -2672,7 +2672,14 @@ void smt2_convt::convert_struct(const struct_exprt &expr)
+
+          if(numeric_cast_v<mp_integer>(to_constant_expr(size_expr)) != 0)
+            flatten_array(op);
+        }
```
In principle, this check would now need to be done in front of every call to flatten array, to make that a precondition of that function.
I think this is the wrong approach. I would let flatten_array handle this case.

NlightNFotis commented:
@kroening Is this something that has value, and we should rework to get in `develop`?

Or has this been sunset on the basis of other changes (I can see there are merge conflicts).

I'm inclined to believe that there's not a lot of enthusiasm for this to get in (given that the last update on this was on 2019) but figured I should ask before I close the PR.

kroening commented:
It certainly makes sense to systematically check that we can do zero-sized arrays with all the solver back-ends.

The PR itself is not quite done yet, and would need a bit of work.
